File: 42-TrappingRainWater/42-TrappingRainWater.cpp
// Last updated: 6/25/2026, 9:17:45 AM
class Solution {
public:
    int trap(vector<int>& height) {
        int left = 0, right = height.size()-1, max_index, max_val = INT_MIN, result = 0, max_left = height[0], max_right = height[height.size()-1];
        for(int i=0; i<height.size(); i++){
            if(max_val < height[i]){
                max_index = i;
                max_val = height[i];
            }
        }
        for(int i=0; i<max_index; i++){
            max_left = max(max_left, height[i]);
            result += max_left - height[i];
        }
        for(int i=height.size()-1; i>max_index; i--){
            max_right = max(max_right, height[i]);
            result += max_right - height[i];
        }
        return result;
    }
};

File: 42-TrappingRainWater/42-TrappingRainWater.py
# Last updated: 6/25/2026, 9:17:40 AM
class Solution(object):
    def trap(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        l,r=0,len(height)-1
        max_left,max_right=height[l],height[r]
        total=0
        while l<=r:
            if max_left<=max_right:
                temp_water=max_left-height[l]
                total+=temp_water if temp_water>0 else 0
                max_left=max(max_left,height[l])
                l+=1
            
            else:
                temp_water=max_right-height[r]
                total+=temp_water if temp_water>0 else 0
                max_right=max(max_right,height[r])
                r-=1
        
        return total
            